hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TypeUtils.java

```diff
@@ -0,0 +1,328 @@
+            RelationsStorage relationsStorage,
+            Map<String, String> options
+    ) {
+        if (!field.type().isCustomType()) {
+            return;
+        }
+        String format = options.get(isKey ? SqlConnector.OPTION_KEY_FORMAT : SqlConnector.OPTION_VALUE_FORMAT);
+        FieldEnricher<?, ?> enricher;
+        switch (format) {
+            case SqlConnector.PORTABLE_FORMAT:
+                enricher = new PortableEnricher(relationsStorage, serializationService);
+                break;
+            case SqlConnector.COMPACT_FORMAT:
+                enricher = new CompactEnricher(relationsStorage);
+                break;
+            case SqlConnector.JAVA_FORMAT:
+                enricher = new JavaEnricher(relationsStorage);
+                break;
+            case SqlConnector.AVRO_FORMAT:
+                enricher = new AvroEnricher(relationsStorage);
+                break;
+            default:
+                throw QueryException.error("Unsupported type format: " + format);
+        }
+        enricher.enrich(field, options, isKey);
+    }
+
+    private static class PortableEnricher extends FieldEnricher<PortableId, ClassDefinition> {
+        private final PortableContext context;
+
+        PortableEnricher(RelationsStorage relationsStorage, InternalSerializationService serializationService) {
+            super(QueryDataType.OBJECT_TYPE_KIND_PORTABLE, relationsStorage);
+            context = serializationService.getPortableContext();
+        }
+
+        @Override
+        protected String getTypeMetadata(PortableId portableId) {
+            return portableId.toString();
+        }
+
+        @Override
+        protected ClassDefinition getSchema(PortableId portableId) {
+            ClassDefinition classDef = context.lookupClassDefinition(portableId);
+            if (classDef == null) {
+                throw QueryException.error("The given factoryId/classId/version combination not known to the member");
+            }
+            return classDef;
+        }
+
+        @Override
+        protected List<TypeField> resolveFields(ClassDefinition classDef) {
+            return classDef.getFieldNames().stream().map(name -> {
+                FieldDefinition field = classDef.getField(name);
+                if (field.getType().equals(FieldType.PORTABLE)) {
+                    throw QueryException.error("Column list is required to create nested fields");
+                }
+                return new TypeField(name, PORTABLE_TO_SQL.getOrDefault(field.getType()));
+            }).collect(toList());
+        }
+
+        @Override
+        protected PortableId getFieldSchemaId(ClassDefinition classDef, String fieldName, String fieldTypeName) {
+            return classDef.getField(fieldName).getPortableId();
+        }
+
+        @Override
+        protected PortableId getSchemaId(Map<String, String> options, Boolean isKey) {
+            return isKey != null ? portableId(options, isKey) : null;
```

Comment:
what happens if the user does not specify necessary serialization format options for given type? We should get clear message, eg. something like "you tried to use UDT <name> with serialization format <xxx> but option <ooo> was not configured". If I read correctly for portable we will get NPE or NumberFormatException.

Reply:
Type metadata is optional and if it is not provided, metadata is inherited from the mapping. If `__key` or `this` is the only mapping field and both the mapping and type specify metadata, the one specified by the type is chosen.